src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -0,0 +1,357 @@
+            switch (operation)
+            {
+                case DataModificationOperationKind.Update:
+                    deltaSetItem = changedObj;
+                    break;
+                case DataModificationOperationKind.Insert:
+                    {
+                        deltaSetItem = CreateDeletedEntityForFailedOperation(changedObj);
+
+                        break;
+                    }
+                case DataModificationOperationKind.Delete:
+                    {
+                        deltaSetItem = CreateEntityObjectforFailedOperation(changedObj, originalObj);                        
+                        break;
+                    }
+            }
+
+
+            deltaSetItem.TransientInstanceAnnotationContainer = changedObj.TransientInstanceAnnotationContainer;
+            deltaSetItem.TransientInstanceAnnotationContainer.AddResourceAnnotation("Core.DataModificationException", dataModificationExceptionType);
+
+            Contract.Assert(deltaSetItem != null);
+
+            return deltaSetItem;
+        }
+
+        private IDeltaSetItem CreateEntityObjectforFailedOperation(Delta<TStructuralType> changedObj, TStructuralType originalObj)
+        {
+            Type type = typeof(Delta<>).MakeGenericType(_clrType);
+
+            Delta<TStructuralType> deltaObject = Activator.CreateInstance(type, _clrType, _clrType.GetProperties().Select(x=>x.Name), null,
+                changedObj.InstanceAnnotationsPropertyInfo) as Delta<TStructuralType>;
+
+            SetProperties(originalObj, deltaObject);
+
+            if (deltaObject.InstanceAnnotationsPropertyInfo != null) {
+
+                object instAnnValue;
+                changedObj.TryGetPropertyValue(deltaObject.InstanceAnnotationsPropertyInfo.Name, out instAnnValue);
+                IODataInstanceAnnotationContainer instanceAnnotations = instAnnValue as IODataInstanceAnnotationContainer;
+
+                if (instanceAnnotations != null)
+                {
+                    deltaObject.TrySetPropertyValue(deltaObject.InstanceAnnotationsPropertyInfo.Name, instanceAnnotations);
+                }
+            }
+
+            return deltaObject;
+        }
+
+        private void SetProperties(TStructuralType originalObj, Delta<TStructuralType> edmDeltaEntityObject)
+        {
+            foreach (string property in edmDeltaEntityObject.GetChangedPropertyNames())
+            {
+                edmDeltaEntityObject.TrySetPropertyValue(property, _clrType.GetProperty(property).GetValue(originalObj));
+            }
+
+            foreach (string property in edmDeltaEntityObject.GetUnchangedPropertyNames())
+            {
+                edmDeltaEntityObject.TrySetPropertyValue(property, _clrType.GetProperty(property).GetValue(originalObj));
+            }
+        }
+
+        private DeltaDeletedEntityObject<TStructuralType> CreateDeletedEntityForFailedOperation(Delta<TStructuralType> changedObj)
```

Comment:
>CreateDeletedEntityForFailedOperation [](start = 58, length = 37)

This has almost exactly the same logic as CreateEntityObjectforFailedOperation.  Can the bulk of the logic (i.e., copying properties and instance annotations) be consolidated into a common method?

Reply:
bcoz logic are diff  In one place we are copying between 2 delta<T>s vs in another case its between a T and delta<T>